internal/task/file_tasks.go

```diff
@@ -0,0 +1,102 @@
+package task
+
+import (
+	"context"
+	"crypto"
+	"fmt"
+
+	"github.com/anchore/syft/internal/sbomsync"
+	"github.com/anchore/syft/syft/artifact"
+	"github.com/anchore/syft/syft/file"
+	"github.com/anchore/syft/syft/file/cataloger/filedigest"
+	"github.com/anchore/syft/syft/file/cataloger/filemetadata"
+	"github.com/anchore/syft/syft/pkg"
+	"github.com/anchore/syft/syft/sbom"
+)
+
+func NewFileDigestCatalogerTask(selection file.Selection, hashers ...crypto.Hash) Task {
+	if selection == file.NoFilesSelection || len(hashers) == 0 {
+		return nil
+	}
+
+	digestsCataloger := filedigest.NewCataloger(hashers)
+
+	fn := func(ctx context.Context, resolver file.Resolver, builder sbomsync.Builder) error {
+		accessor := builder.(sbomsync.Accessor)
+
+		var coordinates []file.Coordinates
+
+		accessor.ReadFromSBOM(func(sbom *sbom.SBOM) {
+			if selection == file.OwnedFilesSelection {
+				for _, r := range sbom.Relationships {
+					// TODO: double check this logic
+					if r.Type != artifact.ContainsRelationship {
+						continue
+					}
+					if _, ok := r.From.(pkg.Package); !ok {
+						continue
+					}
+					if c, ok := r.To.(file.Coordinates); ok {
+						coordinates = append(coordinates, c)
+					}
+				}
+			}
+		})
+
+		result, err := digestsCataloger.Catalog(resolver, coordinates...)
```

Comment:
from @willmurphyscode , blocking: we need to explicitly pass all coordinates, since there is no guarantee to have any results from a owned-files indication

Reply:
I'm going to fix the functional problem in this PR, but to address the signature and generator issue I really should break that into a separate PR that I follow up with after this PR

Reply:
I think this will have some play into the solution here https://github.com/anchore/syft/issues/2487